src/IO/HadoopSnappyReadBuffer.cpp

```diff
@@ -212,6 +212,10 @@ bool HadoopSnappyReadBuffer::nextImpl()
         }
         return true;
     }
+    else if (decoder->result == Status::NEEDS_MORE_INPUT)
```

Comment:
I remember there we similar fix (with recusion) in some another buffer which was l later remade to a while cycle because recursion cause problems. Would it be better to use a cycle here too? Anyway current way is probably also ok.

Reply:
Yep, cycle is better, also thought about it, will rewrite it